libs/geometry/include/geometry/TangentSpaceMesh.h

```diff
@@ -0,0 +1,188 @@
+         * Reference: Frisvad, J.R., 2012. Building an orthonormal basis from a 3D unit vector
+         *     without normalization. Journal of Graphics Tools, 16(3), pp.151-159.
+         *     http://people.compute.dtu.dk/jerf/code/hairy/
+         */
+        FRISVAD = 4,
+
+        /*
+         * --- Flat Shading ---
+         * Requires: positions + indices
+         * Note: Will remesh
+         */
+        FLAT_SHADING = 5,
+
+        /*
+         * --- Sign of W ---
+         * Requires: normals + tangents
+         * Note: The sign of W determines the orientation of the bitangent.
+         */
+        SIGN_OF_W = 6
+    };
+
+    /*
+     * Computation of the tangent space is intended to be synchronous (worked on the same thread).
+     * Client is expected to keep the input immutable and in a good state for the duration of both
+     * computation *and* query. That is, when querying the result of the tangent spaces, part of the
+     * result might depend on the input data.
+     */
+    class Builder {
+    public:
+        Builder() noexcept;
+        ~Builder() noexcept;
+        Builder(Builder&& that) noexcept;
+        Builder& operator=(Builder&& that) noexcept;
+
+        Builder& vertexCount(size_t vertexCount) noexcept;
+
+        Builder& normals(const filament::math::float3*, size_t stride = 0) noexcept;
+        Builder& tangents(const filament::math::float4*, size_t stride = 0) noexcept;
+        Builder& uvs(const filament::math::float2*, size_t stride = 0) noexcept;
+        Builder& positions(const filament::math::float3*, size_t stride = 0) noexcept;
+
+        Builder& triangleCount(size_t triangleCount) noexcept;
+        Builder& triangles(const filament::math::uint3*) noexcept;
+        Builder& triangles(const filament::math::ushort3*) noexcept;
+
+        Builder& algorithmHint(AlgorithmHint) noexcept;
+
+        TangentSpaceMesh* build();
+
+    private:
+        Builder(const Builder&) = delete;
+        Builder& operator=(const Builder&) = delete;
+
+        TangentSpaceMesh* mMesh;
+    };
+
+    ~TangentSpaceMesh() noexcept;
+    TangentSpaceMesh(TangentSpaceMesh&& that) noexcept;
+    TangentSpaceMesh& operator=(TangentSpaceMesh&& that) noexcept;
+
+    size_t getVertexCount() const noexcept;
+
+    /*
+     * The following methods for retrieving vertex-correlated data assumes that the *out* param
+     * provided by the client is at least of *getVertexCount()* length (accounting for stride).
+     *
+     * If the chosen algorithm did not remesh the input, the client is advised to just use the
+     * data they provided instead of querying.  For example, if the chosen method is FRISVAD, then
+     * the client should not need to call "getPositions". We will simply copy from the input
+     * *positions* in that case.
+     *
+     * If the client calls getPositions and positions were not provided as input, we will throw
+     * and exception. Similar behavior will apply to UVs.
+     */
```

Comment:
For public method, use the javadoc style (which we process with doxygen) comments. Most important is to start with `/**` and to document parameters with `@param`. Also use `@see` where appropriate.

Reply:
done